File: ein.py/src/ein/render/rules.py
"""Rule → DOT renderer — S1.6.1.

Renders a single `(rule Name (params…) :match … :assert … :why … …)`
as a DOT digraph showing the rule's *pattern → conclusion* shape. Two
modes:

- **sidebyside** (alias ``"a"``, default) — `match` and `assert` as two
  separate clusters; the most readable view for a rule library.
- **overlay** (alias ``"c"``) — match solid + assert dashed on one
  graph; compact, used inline at rule-firing time.

What this renderer gets right that a naive edge-dump does not
(S1.6.1 T1.6.1.1):

- **Clean labels.** Node ids carry a per-panel suffix so the `match`
  and `assert` copies of a variable don't collapse into one shared
  node, but the suffix never reaches the *label* — both panels show
  `?a`. Variables render as diamonds, ground atoms as rectangles,
  wildcards as dashed diamonds (the shape legend).
- **Constraints are not relations.** A guard predicate `(neq ?a ?b)` /
  `(eq ?a ?b)` is computed, not data — it renders as a dotted,
  undirected ``≠`` / ``=`` link (``constraint=false``), never a
  labelled relation arrow.
- **Negation** `(not (R a b))` renders red with a ``¬`` prefix.
- **NAF guards** `(absent (and …))` render as their own
  ``cluster_absent`` sub-graph — inner conjuncts as arrows *inside* the
  cluster, the binder-local variables declared inside it, shared
  variables left outside — so the "no such match exists" reading
  survives instead of being flattened into overlay arrows.

This is the canonical rule renderer; :mod:`ein.ir.to_dot`
delegates `(rule …)` / `(rules …)` rendering here.
"""
from __future__ import annotations

from ..ir.types import Atom, KwPair, SForm
from .dot_util import (
    GROUND_SHAPE,
    HYPER_SHAPE,
    VAR_SHAPE,
    WILDCARD_ATTRS,
    quote,
    value_label,
)
from .palette import hash_color

# Constraint glyphs for the built-in guard predicates.
_PRED_GLYPH = {"neq": "≠", "eq": "="}
_NEG_COLOUR = "#d62728"      # red — negated premises / conclusions
_GUARD_COLOUR = "#888888"    # grey — NAF guard cluster chrome
_CONSTRAINT_COLOUR = "#555555"

# Rule-mode aliases. Default is side-by-side LHS|RHS clusters (most
# readable for rule libraries); compact overlay is opt-in. Legacy
# single-letter names "a"/"c" stay accepted.
_RULE_MODE_ALIASES = {
    "a": "a", "sidebyside": "a", "side-by-side": "a",
    "c": "c", "overlay": "c",
}


# ── Structural-head classification ─────────────────────────────────

def _is_predicate(name: str) -> bool:
    """Whether a head is a built-in guard predicate (`eq` / `neq`).

    Imported lazily: ``ir.to_dot`` imports this module, and
    ``inference.predicates`` transitively imports ``ir`` — a module-level
    import here would close that cycle.
    """
    from ..inference.predicates import is_predicate
    return is_predicate(name)


def _head_name(expr: SForm) -> str | None:
    """The head atom name, or None when the head is a `?var` / `_`."""
    return expr.head.name if isinstance(expr.head, Atom) else None


def _is_guard(expr: SForm) -> bool:
    return _head_name(expr) in ("absent", "forall")


def _positional(expr: SForm) -> list:
    return [a for a in expr.args if not isinstance(a, KwPair)]


# ── Node-occurrence analysis (for absent-cluster scoping) ──────────

def _arg_nodes(expr: SForm) -> list[str]:
    """Value-labels of all argument positions in this clause tree.

    Relation / predicate *heads* are edge labels, not nodes, so they
    are excluded. Recurses through `and` / `or` / `not` / `absent` /
    `forall` so a guard's inner argument nodes are surfaced.
    """
    hn = _head_name(expr)
    if hn in ("and", "or", "absent", "forall"):
        out: list[str] = []
        for child in expr.args:
            if isinstance(child, SForm):
                out.extend(_arg_nodes(child))
        return out
    if hn == "not":
        if expr.args and isinstance(expr.args[0], SForm):
            return _arg_nodes(expr.args[0])
        return []
    return [value_label(a) for a in _positional(expr)]


def _ordered_nodes(clauses: list[SForm]) -> list[str]:
    """First-seen order of argument-node labels across `clauses`."""
    seen: list[str] = []
    s: set[str] = set()
    for c in clauses:
        for n in _arg_nodes(c):
            if n not in s:
                s.add(n)
                seen.append(n)
    return seen


def _node_homes(top: list[SForm], guards: list[SForm]) -> dict[str, object]:
    """Assign each node a home: ``"top"`` or a guard index.

    A node lives in a guard cluster iff it appears in exactly one guard
    and nowhere in the top-level clauses (it is the guard's binder-local
    variable). Everything shared stays at the top so its edges cross
    into the cluster rather than pulling the node inside.
    """
    top_nodes = {n for c in top for n in _arg_nodes(c)}
    guard_nodes = [set(_arg_nodes(g)) for g in guards]
    homes: dict[str, object] = {}
    every = set(top_nodes)
    for gs in guard_nodes:
        every |= gs
    for n in every:
        in_guards = [i for i, gs in enumerate(guard_nodes) if n in gs]
        homes[n] = in_guards[0] if (n not in top_nodes and len(in_guards) == 1) else "top"
    return homes


# ── Shapes / ids ───────────────────────────────────────────────────

def _shape_attrs(nodelabel: str) -> str:
    if nodelabel == "_":
        return WILDCARD_ATTRS
    if nodelabel.startswith("?"):
        return f"shape={VAR_SHAPE}"
    return f"shape={GROUND_SHAPE}"


def _nid(nodelabel: str, suffix: str) -> str:
    """Quoted DOT id: clean label + a per-panel disambiguating suffix."""
    return quote(nodelabel + suffix)


# ── The renderer ───────────────────────────────────────────────────

class _RuleRenderer:
    """Accumulates the lines for one rule's DOT, panel by panel."""

    def __init__(self) -> None:
        self._hcount = 0

    def _fresh_hyper(self, label: str, suffix: str) -> str:
        self._hcount += 1
        return quote(f"h{self._hcount}_{label}{suffix}")

    # — one clause → lines (relations, constraints, negation, nested) —

    def _clause_lines(self, clause: SForm, *, suffix: str, dashed: bool,
                      negative: bool) -> list[str]:
        hn = _head_name(clause)
        if hn in ("and", "or"):
            out: list[str] = []
            for child in clause.args:
                if isinstance(child, SForm):
                    out.extend(self._clause_lines(
                        child, suffix=suffix, dashed=dashed, negative=negative))
            return out
        if hn == "not":
            inner = clause.args[0] if clause.args else None
            if isinstance(inner, SForm):
                return self._clause_lines(inner, suffix=suffix, dashed=dashed,
                                          negative=True)
            return []
        if hn is not None and _is_predicate(hn):
            return [self._constraint_line(clause, suffix=suffix)]
        # An absent/forall reached here (nested inside another clause) —
        # render its body inline as forbidden (negative); the top-level
        # case is handled as a cluster by `panel`.
        if hn in ("absent", "forall"):
            out = []
            for child in clause.args:
                if isinstance(child, SForm):
                    out.extend(self._clause_lines(child, suffix=suffix,
                                                  dashed=dashed, negative=True))
            return out
        return self._relation_lines(clause, suffix=suffix, dashed=dashed,
                                    negative=negative)

    def _relation_lines(self, clause: SForm, *, suffix: str, dashed: bool,
                        negative: bool) -> list[str]:
        head_label = value_label(clause.head)
        pos = _positional(clause)
        colour = _NEG_COLOUR if negative else hash_color(head_label.lstrip("?"))
        label = ("¬" + head_label) if negative else head_label
        attrs = [f"label={quote(label)}", f'color="{colour}"',
                 f'fontcolor="{colour}"']
        if dashed:
            attrs.append("style=dashed")
        attr_s = ", ".join(attrs)
        if len(pos) == 2:
            a, b = pos
            return [f"{_nid(value_label(a), suffix)} -> "
                    f"{_nid(value_label(b), suffix)} [{attr_s}];"]
        # n-ary (or arity 0/1): a Levi octagon list-node + role edges.
        h = self._fresh_hyper(head_label.lstrip("?"), suffix)
        lines = [f'{h} [shape={HYPER_SHAPE}, label={quote(label)}, '
                 f'color="{colour}", fontcolor="{colour}"];']
        edge_style = ", style=dashed" if dashed else ""
        for i, arg in enumerate(pos, start=1):
            lines.append(f'{h} -> {_nid(value_label(arg), suffix)} '
                         f'[label="{i}", color="{colour}"{edge_style}];')
        return lines

    def _constraint_line(self, clause: SForm, *, suffix: str) -> str:
        glyph = _PRED_GLYPH.get(clause.head.name, clause.head.name)
        pos = _positional(clause)
        if len(pos) != 2:  # defensive — eq/neq are binary
            return (f"// constraint {value_label(clause)}")
        a, b = pos
        return (f"{_nid(value_label(a), suffix)} -> "
                f"{_nid(value_label(b), suffix)} "
                f'[label="{glyph}", dir=none, style=dotted, '
                f'color="{_CONSTRAINT_COLOUR}", '
                f'fontcolor="{_CONSTRAINT_COLOUR}", constraint=false];')

    # — one pattern (match or assert) → (node decls, body lines) —

    def panel(self, pattern: SForm | None, *, suffix: str, dashed: bool,
              ) -> tuple[dict[str, str], list[str]]:
        """Render one pattern into (node-decl map, ordered body lines).

        Top-home nodes are returned in the decl map (the caller emits
        them at panel scope); guard-local nodes are declared inside
        their `cluster_absent` block within the body lines.
        """
        decls: dict[str, str] = {}
        body: list[str] = []
        if pattern is None:
            return decls, body
        hn = _head_name(pattern)
        clauses = ([c for c in pattern.args if isinstance(c, SForm)]
                   if hn == "and" else [pattern])
        top = [c for c in clauses if not _is_guard(c)]
        guards = [c for c in clauses if _is_guard(c)]
        homes = _node_homes(top, guards)

        # Panel-scope node declarations (shared / top-home nodes).
        for nl in _ordered_nodes(clauses):
            if homes.get(nl) == "top":
                decls.setdefault(_nid(nl, suffix),
                                 f"label={quote(nl)}, {_shape_attrs(nl)}")

        # Top-level clause lines.
        for c in top:
            body.extend(self._clause_lines(c, suffix=suffix, dashed=dashed,
                                           negative=False))

        # Each guard → its own cluster (local decls + inner lines).
        for gi, guard in enumerate(guards):
            kind = _head_name(guard) or "absent"
            cid = f"cluster_{kind}_{suffix.strip('_') or 'o'}_{gi}"
            glyph = "∄" if kind == "absent" else "∀"
            block = [
                f"subgraph {cid} {{",
                f'  label="{kind} ({glyph})"; style="dashed,rounded"; '
                f'color="{_GUARD_COLOUR}"; fontcolor="{_GUARD_COLOUR}";',
            ]
            for nl in _ordered_nodes([guard]):
                if homes.get(nl) == gi:
                    block.append(f"  {_nid(nl, suffix)} "
                                 f"[label={quote(nl)}, {_shape_attrs(nl)}];")
            for child in guard.args:
                if isinstance(child, SForm):
                    block.extend(
                        "  " + ln for ln in self._clause_lines(
                            child, suffix=suffix, dashed=dashed, negative=False))
            block.append("}")
            body.extend(block)
        return decls, body


# ── Field extraction ───────────────────────────────────────────────

def _extract(rule_sform: SForm) -> tuple[str, SForm | None, SForm | None]:
    """Return (name, match-expr, assert-expr) from a `(rule …)` form."""
    name = "anon"
    match_expr: SForm | None = None
    assert_expr: SForm | None = None
    for i, arg in enumerate(rule_sform.args):
        if i == 0 and isinstance(arg, Atom):
            name = arg.name
        elif isinstance(arg, KwPair):
            if arg.key.name == "match" and isinstance(arg.value, SForm):
                match_expr = arg.value
            elif arg.key.name == "assert" and isinstance(arg.value, SForm):
                assert_expr = arg.value
    return name, match_expr, assert_expr


# ── Public entry points ────────────────────────────────────────────

def render_rule(rule_sform: SForm, *, mode: str = "a") -> str:
    """Render one `(rule …)` form as a DOT digraph string.

    ``mode``: ``"sidebyside"`` / ``"a"`` (default) or ``"overlay"`` /
    ``"c"``. Legacy single-letter names are accepted.
    """
    canon = _RULE_MODE_ALIASES.get(mode)
    if canon is None:
        raise ValueError(
            f"unknown rule mode: {mode!r} "
            f"(expected 'sidebyside'/'a' or 'overlay'/'c')"
        )
    name, match_expr, assert_expr = _extract(rule_sform)
    safe = name.replace("-", "_").replace(" ", "_")
    if canon == "a":
        return _render_sidebyside(safe, name, match_expr, assert_expr)
    return _render_overlay(safe, name, match_expr, assert_expr)


def _render_sidebyside(safe: str, name: str, match: SForm | None,
                       assert_: SForm | None) -> str:
    r = _RuleRenderer()
    m_decls, m_body = r.panel(match, suffix="_L", dashed=False)
    a_decls, a_body = r.panel(assert_, suffix="_R", dashed=True)
    out = [f"digraph rule_{safe}_lhs_rhs {{",
           "  rankdir=TB;",
           f"  label={quote(name)}; labelloc=t;",
           '  subgraph cluster_lhs { label="match";']
    out += [f"    {ln}" for ln in _decl_lines(m_decls)]
    out += [f"    {ln}" for ln in m_body]
    out.append("  }")
    out.append('  subgraph cluster_rhs { label="assert";')
    out += [f"    {ln}" for ln in _decl_lines(a_decls)]
    out += [f"    {ln}" for ln in a_body]
    out.append("  }")
    out.append("}")
    return "\n".join(out)


def _render_overlay(safe: str, name: str, match: SForm | None,
                    assert_: SForm | None) -> str:
    r = _RuleRenderer()
    m_decls, m_body = r.panel(match, suffix="", dashed=False)
    a_decls, a_body = r.panel(assert_, suffix="", dashed=True)
    merged = dict(m_decls)
    for k, v in a_decls.items():
        merged.setdefault(k, v)
    out = [f"digraph rule_{safe}_overlay {{",
           f"  label={quote(name)}; labelloc=t;"]
    out += [f"  {ln}" for ln in _decl_lines(merged)]
    out += [f"  {ln}" for ln in m_body]
    out += [f"  {ln}" for ln in a_body]
    out.append("}")
    return "\n".join(out)


def _decl_lines(decls: dict[str, str]) -> list[str]:
    return [f"{nid} [{attrs}];" for nid, attrs in decls.items()]


def render_rules(rules_sform: SForm, *, mode: str = "a") -> str:
    """Render a `(rules …)` library — one digraph per child rule."""
    chunks = [render_rule(r, mode=mode)
              for r in rules_sform.args if isinstance(r, SForm)]
    return "\n\n".join(chunks)


__all__ = ["render_rule", "render_rules"]

File: ein.rs/crates/ein-render/src/rules.rs
//! The rule → DOT renderer — `ein.py`'s `render/rules.py`.
//!
//! A `(rule Name (params…) :match … :assert … …)` as its *pattern →
//! conclusion* shape, in two modes: **sidebyside** (two clusters, the
//! readable view for a rule library) and **overlay** (match solid, assert
//! dashed, on one graph — the compact inline view).
//!
//! What it gets right that an edge dump does not: a per-panel id suffix that
//! never reaches the *label*, so `?a` appears as `?a` in both panels without
//! the two copies collapsing; guard predicates drawn as dotted undirected
//! `≠` / `=` links rather than as relation arrows, because they are computed
//! and not data; negation in red with a `¬` prefix; and a NAF `absent` guard
//! as its own `cluster_absent`, with the guard's binder-local variables
//! declared *inside* it and shared variables left outside — so "no such match
//! exists" survives instead of flattening into overlay arrows.

use ein_core::is_predicate;
use ein_ir::{Ast, Node, NodeId};

use crate::dot_util::{GROUND_SHAPE, HYPER_SHAPE, VAR_SHAPE, WILDCARD_ATTRS, quote, value_label};
use crate::palette::hash_color;

const NEG_COLOUR: &str = "#d62728"; // red — negated premises / conclusions
const GUARD_COLOUR: &str = "#888888"; // grey — NAF guard cluster chrome
const CONSTRAINT_COLOUR: &str = "#555555";

/// Constraint glyphs for the built-in guard predicates.
fn pred_glyph(name: &str) -> &str {
    match name {
        "neq" => "≠",
        "eq" => "=",
        other => other,
    }
}

/// Side-by-side LHS|RHS clusters (default) or the compact overlay. The legacy
/// single-letter names stay accepted.
#[derive(Clone, Copy, PartialEq, Eq)]
pub enum RuleMode {
    SideBySide,
    Overlay,
}

impl RuleMode {
    pub fn parse(s: &str) -> Option<RuleMode> {
        match s {
            "a" | "sidebyside" | "side-by-side" => Some(RuleMode::SideBySide),
            "c" | "overlay" => Some(RuleMode::Overlay),
            _ => None,
        }
    }
}

// ── structural-head classification ─────────────────────────────────

/// The head atom's name, or `None` when the head is a `?var` / `_`.
fn head_name(ast: &Ast, expr: NodeId) -> Option<&str> {
    match ast.node(expr) {
        Node::SForm { head, .. } => match ast.node(head) {
            Node::Atom(s) => Some(ast.sym(s)),
            _ => None,
        },
        _ => None,
    }
}

fn is_guard(ast: &Ast, expr: NodeId) -> bool {
    matches!(head_name(ast, expr), Some("absent" | "forall"))
}

fn positional(ast: &Ast, expr: NodeId) -> Vec<NodeId> {
    ast.form_args(expr)
        .iter()
        .copied()
        .filter(|a| !matches!(ast.node(*a), Node::KwPair { .. }))
        .collect()
}

fn form_head(ast: &Ast, expr: NodeId) -> NodeId {
    match ast.node(expr) {
        Node::SForm { head, .. } => head,
        _ => expr,
    }
}

fn is_sform(ast: &Ast, id: NodeId) -> bool {
    matches!(ast.node(id), Node::SForm { .. })
}

// ── node-occurrence analysis (for absent-cluster scoping) ──────────

/// Value-labels of every argument position in this clause tree.
///
/// Relation and predicate *heads* are edge labels, not nodes, so they are
/// excluded; the walk recurses through `and` / `or` / `not` / `absent` /
/// `forall` so a guard's inner argument nodes surface.
fn arg_nodes(ast: &Ast, expr: NodeId) -> Vec<String> {
    match head_name(ast, expr) {
        Some("and" | "or" | "absent" | "forall") => {
            let mut out = Vec::new();
            for child in ast.form_args(expr).to_vec() {
                if is_sform(ast, child) {
                    out.extend(arg_nodes(ast, child));
                }
            }
            out
        }
        Some("not") => match ast.form_args(expr).first().copied() {
            Some(first) if is_sform(ast, first) => arg_nodes(ast, first),
            _ => Vec::new(),
        },
        _ => positional(ast, expr)
            .iter()
            .map(|a| value_label(ast, *a))
            .collect(),
    }
}

/// First-seen order of argument-node labels across `clauses`.
fn ordered_nodes(ast: &Ast, clauses: &[NodeId]) -> Vec<String> {
    let mut seen: Vec<String> = Vec::new();
    for c in clauses {
        for n in arg_nodes(ast, *c) {
            if !seen.contains(&n) {
                seen.push(n);
            }
        }
    }
    seen
}

/// Where a node lives: at panel scope, or inside guard `n`'s cluster.
#[derive(Clone, Copy, PartialEq, Eq)]
enum Home {
    Top,
    Guard(usize),
}

/// A node lives in a guard cluster iff it appears in exactly one guard and
/// nowhere in the top-level clauses — that makes it the guard's binder-local
/// variable. Everything shared stays at the top so its edges cross into the
/// cluster rather than pulling the node inside.
fn node_homes(ast: &Ast, top: &[NodeId], guards: &[NodeId]) -> Vec<(String, Home)> {
    let mut top_nodes: Vec<String> = Vec::new();
    for c in top {
        for n in arg_nodes(ast, *c) {
            if !top_nodes.contains(&n) {
                top_nodes.push(n);
            }
        }
    }
    let guard_nodes: Vec<Vec<String>> = guards.iter().map(|g| arg_nodes(ast, *g)).collect();
    let mut every: Vec<String> = top_nodes.clone();
    for gs in &guard_nodes {
        for n in gs {
            if !every.contains(n) {
                every.push(n.clone());
            }
        }
    }
    every
        .into_iter()
        .map(|n| {
            let in_guards: Vec<usize> = guard_nodes
                .iter()
                .enumerate()
                .filter(|(_, gs)| gs.contains(&n))
                .map(|(i, _)| i)
                .collect();
            let home = if !top_nodes.contains(&n) && in_guards.len() == 1 {
                Home::Guard(in_guards[0])
            } else {
                Home::Top
            };
            (n, home)
        })
        .collect()
}

fn home_of(homes: &[(String, Home)], label: &str) -> Option<Home> {
    homes.iter().find(|(n, _)| n == label).map(|(_, h)| *h)
}

// ── shapes / ids ───────────────────────────────────────────────────

fn shape_attrs(nodelabel: &str) -> String {
    if nodelabel == "_" {
        WILDCARD_ATTRS.to_string()
    } else if nodelabel.starts_with('?') {
        format!("shape={VAR_SHAPE}")
    } else {
        format!("shape={GROUND_SHAPE}")
    }
}

/// Quoted DOT id: the clean label plus a per-panel disambiguating suffix.
fn nid(nodelabel: &str, suffix: &str) -> String {
    quote(&format!("{nodelabel}{suffix}"))
}

// ── the renderer ───────────────────────────────────────────────────

/// Accumulates one rule's DOT, panel by panel. The hyperedge counter is
/// per-*rule*, not per-panel, so an n-ary clause on the right keeps counting
/// from where the left panel stopped.
#[derive(Default)]
struct RuleRenderer {
    hcount: u32,
}

/// A panel's node declarations, in first-declaration order.
type Decls = Vec<(String, String)>;

fn decls_setdefault(decls: &mut Decls, key: String, value: String) {
    if !decls.iter().any(|(k, _)| *k == key) {
        decls.push((key, value));
    }
}

impl RuleRenderer {
    fn fresh_hyper(&mut self, label: &str, suffix: &str) -> String {
        self.hcount += 1;
        quote(&format!("h{}_{label}{suffix}", self.hcount))
    }

    fn clause_lines(
        &mut self,
        ast: &Ast,
        clause: NodeId,
        suffix: &str,
        dashed: bool,
        negative: bool,
    ) -> Vec<String> {
        match head_name(ast, clause) {
            Some("and" | "or") => {
                let mut out = Vec::new();
                for child in ast.form_args(clause).to_vec() {
                    if is_sform(ast, child) {
                        out.extend(self.clause_lines(ast, child, suffix, dashed, negative));
                    }
                }
                out
            }
            Some("not") => match ast.form_args(clause).first().copied() {
                Some(inner) if is_sform(ast, inner) => {
                    self.clause_lines(ast, inner, suffix, dashed, true)
                }
                _ => Vec::new(),
            },
            Some(hn) if is_predicate(hn) => vec![self.constraint_line(ast, clause, suffix)],
            // An `absent` / `forall` reached here is nested inside another
            // clause — render its body inline as forbidden. The top-level
            // case is a cluster, handled by `panel`.
            Some("absent" | "forall") => {
                let mut out = Vec::new();
                for child in ast.form_args(clause).to_vec() {
                    if is_sform(ast, child) {
                        out.extend(self.clause_lines(ast, child, suffix, dashed, true));
                    }
                }
                out
            }
            _ => self.relation_lines(ast, clause, suffix, dashed, negative),
        }
    }

    fn relation_lines(
        &mut self,
        ast: &Ast,
        clause: NodeId,
        suffix: &str,
        dashed: bool,
        negative: bool,
    ) -> Vec<String> {
        let head_label = value_label(ast, form_head(ast, clause));
        let pos = positional(ast, clause);
        let colour = if negative {
            NEG_COLOUR
        } else {
            hash_color(head_label.trim_start_matches('?'))
        };
        let label = if negative {
            format!("¬{head_label}")
        } else {
            head_label.clone()
        };
        let mut attrs = vec![
            format!("label={}", quote(&label)),
            format!("color=\"{colour}\""),
            format!("fontcolor=\"{colour}\""),
        ];
        if dashed {
            attrs.push("style=dashed".to_string());
        }
        let attr_s = attrs.join(", ");
        if pos.len() == 2 {
            return vec![format!(
                "{} -> {} [{attr_s}];",
                nid(&value_label(ast, pos[0]), suffix),
                nid(&value_label(ast, pos[1]), suffix)
            )];
        }
        // n-ary (or arity 0/1): a Levi octagon list-node + role edges.
        let h = self.fresh_hyper(head_label.trim_start_matches('?'), suffix);
        let mut lines = vec![format!(
            "{h} [shape={HYPER_SHAPE}, label={}, color=\"{colour}\", fontcolor=\"{colour}\"];",
            quote(&label)
        )];
        let edge_style = if dashed { ", style=dashed" } else { "" };
        for (i, arg) in pos.iter().enumerate() {
            lines.push(format!(
                "{h} -> {} [label=\"{}\", color=\"{colour}\"{edge_style}];",
                nid(&value_label(ast, *arg), suffix),
                i + 1
            ));
        }
        lines
    }

    fn constraint_line(&mut self, ast: &Ast, clause: NodeId, suffix: &str) -> String {
        let head = form_head(ast, clause);
        let head_name = match ast.node(head) {
            Node::Atom(s) => ast.sym(s).to_string(),
            _ => value_label(ast, head),
        };
        let glyph = pred_glyph(&head_name).to_string();
        let pos = positional(ast, clause);
        if pos.len() != 2 {
            // Defensive — `eq` / `neq` are binary.
            return format!("// constraint {}", value_label(ast, clause));
        }
        format!(
            "{} -> {} [label=\"{glyph}\", dir=none, style=dotted, \
             color=\"{CONSTRAINT_COLOUR}\", fontcolor=\"{CONSTRAINT_COLOUR}\", constraint=false];",
            nid(&value_label(ast, pos[0]), suffix),
            nid(&value_label(ast, pos[1]), suffix)
        )
    }

    /// One pattern → (node-decl map, ordered body lines).
    ///
    /// Top-home nodes come back in the decl map, for the caller to emit at
    /// panel scope; guard-local nodes are declared inside their
    /// `cluster_absent` block within the body lines.
    fn panel(
        &mut self,
        ast: &Ast,
        pattern: Option<NodeId>,
        suffix: &str,
        dashed: bool,
    ) -> (Decls, Vec<String>) {
        let mut decls: Decls = Vec::new();
        let mut body: Vec<String> = Vec::new();
        let Some(pattern) = pattern else {
            return (decls, body);
        };
        let clauses: Vec<NodeId> = if head_name(ast, pattern) == Some("and") {
            ast.form_args(pattern)
                .iter()
                .copied()
                .filter(|c| is_sform(ast, *c))
                .collect()
        } else {
            vec![pattern]
        };
        let top: Vec<NodeId> = clauses
            .iter()
            .copied()
            .filter(|c| !is_guard(ast, *c))
            .collect();
        let guards: Vec<NodeId> = clauses
            .iter()
            .copied()
            .filter(|c| is_guard(ast, *c))
            .collect();
        let homes = node_homes(ast, &top, &guards);

        // Panel-scope node declarations (shared / top-home nodes).
        for nl in ordered_nodes(ast, &clauses) {
            if home_of(&homes, &nl) == Some(Home::Top) {
                decls_setdefault(
                    &mut decls,
                    nid(&nl, suffix),
                    format!("label={}, {}", quote(&nl), shape_attrs(&nl)),
                );
            }
        }

        for c in &top {
            body.extend(self.clause_lines(ast, *c, suffix, dashed, false));
        }

        // Each guard → its own cluster (local decls + inner lines).
        for (gi, guard) in guards.iter().enumerate() {
            let kind = head_name(ast, *guard).unwrap_or("absent").to_string();
            let scope = suffix.trim_matches('_');
            let scope = if scope.is_empty() { "o" } else { scope };
            let cid = format!("cluster_{kind}_{scope}_{gi}");
            let glyph = if kind == "absent" { "∄" } else { "∀" };
            let mut block = vec![
                format!("subgraph {cid} {{"),
                format!(
                    "  label=\"{kind} ({glyph})\"; style=\"dashed,rounded\"; \
                     color=\"{GUARD_COLOUR}\"; fontcolor=\"{GUARD_COLOUR}\";"
                ),
            ];
            for nl in ordered_nodes(ast, &[*guard]) {
                if home_of(&homes, &nl) == Some(Home::Guard(gi)) {
                    block.push(format!(
                        "  {} [label={}, {}];",
                        nid(&nl, suffix),
                        quote(&nl),
                        shape_attrs(&nl)
                    ));
                }
            }
            for child in ast.form_args(*guard).to_vec() {
                if is_sform(ast, child) {
                    for ln in self.clause_lines(ast, child, suffix, dashed, false) {
                        block.push(format!("  {ln}"));
                    }
                }
            }
            block.push("}".to_string());
            body.extend(block);
        }
        (decls, body)
    }
}

// ── field extraction ───────────────────────────────────────────────

/// `(name, match-expr, assert-expr)` from a `(rule …)` form.
fn extract(ast: &Ast, rule: NodeId) -> (String, Option<NodeId>, Option<NodeId>) {
    let mut name = "anon".to_string();
    let mut match_expr = None;
    let mut assert_expr = None;
    for (i, arg) in ast.form_args(rule).to_vec().into_iter().enumerate() {
        match ast.node(arg) {
            Node::Atom(s) if i == 0 => name = ast.sym(s).to_string(),
            Node::KwPair { key, value } => {
                let key = match ast.node(key) {
                    Node::Keyword(s) => ast.sym(s),
                    _ => continue,
                };
                if key == "match" && is_sform(ast, value) {
                    match_expr = Some(value);
                } else if key == "assert" && is_sform(ast, value) {
                    assert_expr = Some(value);
                }
            }
            _ => {}
        }
    }
    (name, match_expr, assert_expr)
}

// ── public entry points ────────────────────────────────────────────

/// One `(rule …)` form as a DOT digraph.
pub fn render_rule_form(ast: &Ast, rule: NodeId, mode: RuleMode) -> String {
    let (name, match_expr, assert_expr) = extract(ast, rule);
    let safe = name.replace(['-', ' '], "_");
    match mode {
        RuleMode::SideBySide => render_sidebyside(ast, &safe, &name, match_expr, assert_expr),
        RuleMode::Overlay => render_overlay(ast, &safe, &name, match_expr, assert_expr),
    }
}

fn decl_lines(decls: &Decls) -> Vec<String> {
    decls
        .iter()
        .map(|(nid, attrs)| format!("{nid} [{attrs}];"))
        .collect()
}

fn render_sidebyside(
    ast: &Ast,
    safe: &str,
    name: &str,
    match_expr: Option<NodeId>,
    assert_expr: Option<NodeId>,
) -> String {
    let mut r = RuleRenderer::default();
    let (m_decls, m_body) = r.panel(ast, match_expr, "_L", false);
    let (a_decls, a_body) = r.panel(ast, assert_expr, "_R", true);
    let mut out = vec![
        format!("digraph rule_{safe}_lhs_rhs {{"),
        "  rankdir=TB;".to_string(),
        format!("  label={}; labelloc=t;", quote(name)),
        "  subgraph cluster_lhs { label=\"match\";".to_string(),
    ];
    out.extend(decl_lines(&m_decls).iter().map(|ln| format!("    {ln}")));
    out.extend(m_body.iter().map(|ln| format!("    {ln}")));
    out.push("  }".to_string());
    out.push("  subgraph cluster_rhs { label=\"assert\";".to_string());
    out.extend(decl_lines(&a_decls).iter().map(|ln| format!("    {ln}")));
    out.extend(a_body.iter().map(|ln| format!("    {ln}")));
    out.push("  }".to_string());
    out.push("}".to_string());
    out.join("\n")
}

fn render_overlay(
    ast: &Ast,
    safe: &str,
    name: &str,
    match_expr: Option<NodeId>,
    assert_expr: Option<NodeId>,
) -> String {
    let mut r = RuleRenderer::default();
    let (m_decls, m_body) = r.panel(ast, match_expr, "", false);
    let (a_decls, a_body) = r.panel(ast, assert_expr, "", true);
    let mut merged = m_decls.clone();
    for (k, v) in a_decls {
        decls_setdefault(&mut merged, k, v);
    }
    let mut out = vec![
        format!("digraph rule_{safe}_overlay {{"),
        format!("  label={}; labelloc=t;", quote(name)),
    ];
    out.extend(decl_lines(&merged).iter().map(|ln| format!("  {ln}")));
    out.extend(m_body.iter().map(|ln| format!("  {ln}")));
    out.extend(a_body.iter().map(|ln| format!("  {ln}")));
    out.push("}".to_string());
    out.join("\n")
}

/// A rule library — one digraph per child rule, joined by blank lines.
pub fn render_rules_forms(ast: &Ast, rules: &[NodeId], mode: RuleMode) -> String {
    let chunks: Vec<String> = rules
        .iter()
        .filter(|r| is_sform(ast, **r))
        .map(|r| render_rule_form(ast, *r, mode))
        .collect();
    chunks.join("\n\n")
}
